### ISSUE TITLE
[BUILD] -std=c++1? compiler flag v.s. non-standard alloca(3) usage

### ISSUE THREAD

**Issue by obache** [OpenImageIO/oiio#2366]
**Describe the bug**
`-std=c++11` compiler flag is set in src/cmake/compiler.cmake,
whereas alloca(3) is used in various source codes.

For alloca(3), compiler built-in implementation should be used, but,
if `-std=` flag is set and forced to use certain standard language specification,
such non-standard buint-in `alloca()` may be hidden.
At least, it is happend on NetBSD with GCC.

**To Reproduce**
1. cmake
2. make
3. error in the link phase, "Warning: reference to the libc supplied alloca(3); this most likely will not work. Please use the compiler privided version of alloca(3), by supplying the appropriate compiler flags (e.g. not -std=c89).
  
**Expected behavior**
Not such warnings, GCC builtin alloca(3) is used.

**Evidence**
If `-std=gnu++11` is used instead of `-std=c++11`, or include `<stdlib.h>` before `alloca(3)` usage, such warning will be disappeared.

**Platform information:**
 - OIIO branch/version: 1.8.15
 - OS: NetBSD 8.1
 - C++ compiler: gcc (nb3 20180905) 5.5.0
 - Any non-default build flags when you build OIIO:

**Comment by lgritz** [comment]
Proposed https://github.com/OpenImageIO/oiio/pull/2367 as a way to fix and clean this up. This is just a patch for master branch. I might backport just the `include` to RB-2.0.

**Comment by obache** [comment]
Yes, it resolve the issue absolutely!

In src/include/OpenImageIO/platform.h and src/include/OpenImageIO/sysutil.h
~~~ c
#ifdef __MINGW32__
#    include <malloc.h>  // for alloca
#endif
~~~
It might the approach to the issue on Windows.

**Comment by obache** [comment]
My falut :( , remaining `-std=gnu++11` flag for testing.
`#include <cstdlib>` only declare C standards, not including `alloca`.

**Comment by lgritz** [comment]
The NetBSD man pages (https://netbsd.gw.com/cgi-bin/man-cgi?alloca+3+NetBSD-current) seems to indicate that stdlib is the right header to include. If this is not correct, do you know which one is? I'd like to avoid `-std=gnu++11` because I don't know what other problems that will cause down the road, it's nonstandard.

**Comment by obache** [comment]
It is right header for `alloca`, but comiler builtin `alloca` is disabled if `-std=c++11`, then libc compat `alloca` symbol will be picked up.

- rewrite nonstandard `alloca` usage with C++ standard way --- right way but troublesome
- use `-std=gnu++11` nonstandard flag if compiler accept it because `alloca` is not in standard -- maybe side effect as you said
- treat NetBSD as a special -- minimal changes

example minimal code changes for the last case (MinGW case already exists):
~~~ c
// for alloca
#ifdef __MINGW32__
#    include <malloc.h>
#elif defined(__NetBSD__) && defined(__GNUC__)
#  ifndef alloca
#    define alloca __builtin_alloca
#  endif
#endif
~~~
 
Other platforms may require another treatments, for example, it seems that AIX with IBM XL C/C++ require `#pragma alloca`:
https://www.ibm.com/support/knowledgecenter/ssw_aix_72/m_bostechref/malloc.html#malloc__alloca

**Comment by lgritz** [comment]
Amended #2367 to include this logic (more or less -- I use __builtin_alloca for all gcc now, which seems even safer).

**Comment by lgritz** [comment]
This was fixed by #2367.
